library/char-stream.lisp

```diff
@@ -0,0 +1,818 @@
+   #:Stream
+   #:open?
+   #:close!
+
+   ;; Input class and its methods
+   #:Input
+   #:read-char!
+   #:read-line!
+
+   ;; additional Input non-method operations
+   #:input-chars!
+   #:input-lines!
+
+   ;; flush operations on Output streams
+   #:FlushOperation
+   #:FlushBlocking
+   #:FlushAsync
+
+   ;; Output class and its methods
+   #:Output
+   #:write-char!
+   #:write-string!
+   #:flush-output!
+
+   ;; additional Output non-method operations
+   #:newline!
+   #:write-line!
+   #:write-chars!
+   #:write-lines!
+   #:write-strings!
+   #:finish-output!
+   #:force-output!
+
+   ;; opening files
+
+   ;; paths
+   #:Path
+
+   ;; encodings
+   #:Encoding
+   #:ASCII
+   #:UTF-8
+   #:UTF-16
+   #:LATIN-1
+   #:default-encoding
+
+   ;; what to do if opening a file that already exists
+   #:IfExists
+   #:IfExistsError
+   #:IfExistsRename
+   #:IfExistsAppend
+   #:IfExistsOverwrite
+   #:IfExistsSupersede
+
+   ;; what to do if opening a file that does not exist
+   #:IfDoesNotExist
+   #:IfDoesNotExistError
+   #:IfDoesNotExistCreate
+
+   ;; options for opening files, incl. encoding, if-exists and if-does-not-exist
+   #:FileOptions
+   #:default-file-options
+   #:with-encoding
+   #:if-exists
+   #:if-does-not-exist
+
+   ;; errors while opening files
+   #:FileError
+
+   ;; actually opening files
+   #:open-input-file!
+   #:open-output-file!
+   #:open-two-way-file!
+
+   ;; passing streams to lisp
+   #:wrap-input-stream-for-lisp
+   #:wrap-output-stream-for-lisp
+   #:wrap-two-way-stream-for-lisp
+
+   ;; common lisp condition classes
+   #:coalton-stream-error
+   #:coalton-stream-error-simple
+   #:coalton-stream-error-simple-message
+   #:coalton-stream-error-closed
+   #:coalton-stream-error-end-of-file
+   #:coalton-stream-error-decoding
+   #:coalton-stream-error-encoding
+   #:coalton-stream-error-timeout
+   #:coalton-stream-error-reader))
+
+#+coalton-release
+(cl:declaim #.coalton-impl:*coalton-optimize-library*)
+
+(cl:in-package #:coalton-library/char-stream)
+
+(cl:defmacro %define-into-as-unsafe-coerce (from-type to-type)
+  (alexandria:with-gensyms (instance)
+    `(define-instance (Into ,from-type ,to-type)
+       (define (into ,instance)
+         (lisp ,to-type (,instance)
+           ,instance)))))
+
+(coalton-toplevel
+  ;; users should interact with streams via the classes `Input' and `Output', rather than explicitly
+  ;; referencing these types
+  (repr :native cl:stream)
+  (define-type %AbstractStream
+    "A Common Lisp `cl:stream' object, with no claims made as to its direction.
+
+Users should interact with streams via the `Stream' class, rather than explicitly referencing this type.")
+
+  ;; it would be nice if this type could be `(and stream (satisfies input-stream-p))', but on SBCL
+  ;; `input-stream-p' and `output-stream-p' return nil for closed input- and output-streams, so that breaks
+  ;; after you close a stream.
+  (repr :native cl:stream)
+  (define-type %InputStream
```

Comment:
Export `%InputStream` `%OutputStream` and `%TwoWayStream`. Un-exported types won't show up on the doc generator.

Reply:
I don't really feel good about exporting types that should not be referred to by user code... would it be acceptable if instead I added a note to the docstring of the `open-*-file` functions that they return a concrete type which should not be named by user code, and stuck in an example for how user code can interact with streams using type variables and class constraints?